drud/ddev
# Add support for all `composer self-update` version formats

## The Problem/Issue/Bug:

The preinstalled Composer in the prebuilt base image is very fast out of date and therefor the latest features of the self-update command are not available.

The PR https://github.com/drud/ddev/pull/3808 was the first step to solve long standing issues regarding the installed Composer version. But as we can see it goes not far enough e.g. Composer supports now also to select minor versions which currently fails because not the latest version is installed in the container. This can be tested my adding the `composer_version: "--2.2"` to the `config.yaml` which currently fails also on master.

## How this PR Solves The Problem:

Before updating to the specified version by the user the latest stable release is always installed. This makes sure latest features can be properly used like the newly introduced minor version which is available in the latest Composer releases on various branches.

In the same time also the predefined key words like `stable`, `preview` and `snapshot` are now supported so the user has now the same options like with manually run `composer self-update`. This makes the Composer version selection in DDEV more streamlined to Composer itself and also others tools like https://github.com/shivammathur/setup-php which also fully supports the various versions formats supported by Composer. So DDEV not longer uses an own strategy here but is more compatible with other tools which should it make easier for the users in the end.

The versioning scheme used before is still fully supported, so no breaking change for existing project but new options where needed.

## Manual Testing Instructions:

- Add an invalid version to the `config.yaml` e.g. `invalid` and check if latest stable version is installed afterwards
- Add a minor version e.g. `2.2` and check if the latest version of this branch is installed afterwards
- Add a version key word like `snapshot` and check if the current master is properly installed

## Automated Testing Overview:

The existing test is enhanced to also test major and minor versions support. The version key words are also used in the iteration but currently I see no easy way to also properly test if the correct version was installed. Added a TODO therefor to enhance these tests too at some time.

## Related Issue Link(s):

## Release/Deployment notes:




<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/3812"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
